### PR TITLE
SPIN1305: Add custom date range queries

### PR DESCRIPTION
**Pull request by damnski**
* Add custom date range queries to cost explorer API wrapper
* Enable cache to allow multiple different date ranges.

### CODE REVIEW

File: api/handlers_spaces.go [hunk 1]
```diff
@@ -40,12 +45,36 @@ func (s *server) SpaceGetHandler(w http.ResponseWriter, r *http.Request) {
...
+		}
+
+		endTime = fmt.Sprintf("%d-%02d-%02d", y, m, d)
+		startTime = fmt.Sprintf("%d-%02d-01", y, m)
+
+	} else {
+		sTime, err := time.Parse("1066-01-01", startTime)
+		if err != nil {
+			log.Infof("error parsing startTime: %s", err)
+			// do we do something more interesting here?  Throw an HTTP code API , such as ErrBadRequest
```
Comment (tenyo) [inline api/handlers_spaces.go]: yes, I think we should `handleError(w, apierror.New(apierror.ErrBadRequest, msg, nil))`

File: api/handlers_spaces.go [hunk 2]
```diff
@@ -40,12 +45,36 @@ func (s *server) SpaceGetHandler(w http.ResponseWriter, r *http.Request) {
...
+		}
+		eTime, err := time.Parse("1066-01-01", endTime)
+		if err != nil {
+			log.Infof("error parsing endTime: %s", err)
+			// do we do something more interesting here?  Throw an HTTP code API , such as ErrBadRequest
```
Comment (tenyo) [inline api/handlers_spaces.go]: same

File: api/handlers_spaces.go [hunk 3]
```diff
@@ -40,12 +45,36 @@ func (s *server) SpaceGetHandler(w http.ResponseWriter, r *http.Request) {
...
+		y, m, d := time.Now().Date()
+		if d == 1 {
+			d = 2
+		}
+
+		endTime = fmt.Sprintf("%d-%02d-%02d", y, m, d)
+		startTime = fmt.Sprintf("%d-%02d-01", y, m)
+
+	} else {
+		sTime, err := time.Parse("1066-01-01", startTime)
```
Comment (tenyo) [inline api/handlers_spaces.go]: Is 1066 the standard format for a year or it doesn't matter?

File: api/routes.go [hunk 4]
```diff
@@ -12,5 +12,9 @@ func (s *server) routes() {
 	api.HandleFunc("/version", s.VersionHandler).Methods(http.MethodGet)
 	api.Handle("/metrics", promhttp.Handler()).Methods(http.MethodGet)
 
+	api.HandleFunc("/{account}/spaces/{space}", s.SpaceGetHandler).
+		Queries("EndTime", "{EndTime}", "StartTime", "{StartTime}").
+		Methods(http.MethodGet)
 	api.HandleFunc("/{account}/spaces/{space}", s.SpaceGetHandler).Methods(http.MethodGet)
+
```
Comment (tenyo) [inline api/routes.go]: ✂️

File: api/handlers_spaces.go [hunk 5]
```diff
@@ -40,12 +45,36 @@ func (s *server) SpaceGetHandler(w http.ResponseWriter, r *http.Request) {
...
+	if endTime == "" || startTime == "" {
+		// if it's the first day of the month, get today's usage thus far
+		y, m, d := time.Now().Date()
+		if d == 1 {
+			d = 2
+		}
+
+		endTime = fmt.Sprintf("%d-%02d-%02d", y, m, d)
+		startTime = fmt.Sprintf("%d-%02d-01", y, m)
+
```
Comment (tenyo) [inline api/handlers_spaces.go]: ✂️

File: api/handlers_spaces.go [hunk 6]
```diff
@@ -147,4 +183,7 @@ func (s *server) SpaceGetHandler(w http.ResponseWriter, r *http.Request) {
 	w.Header().Set("Content-Type", "application/json")
 	w.WriteHeader(http.StatusOK)
 	w.Write(j)
+
+	log.Debugf("print endTime: %s", endTime)
+	log.Debugf("print startTime: %s", startTime)
```
Comment (tenyo) [inline api/handlers_spaces.go]: should remove these before merging

File: api/handlers_spaces.go (hunk 1 above)
Comment (damnski) [reply api/handlers_spaces.go]: done

File: api/handlers_spaces.go (hunk 2 above)
Comment (damnski) [reply api/handlers_spaces.go]: and good.

File: api/handlers_spaces.go (hunk 3 above)
Comment (damnski) [reply api/handlers_spaces.go]: ancient history - year of norman invasion.   I replaced that parse with the "special layout parameter":
https://golang.org/pkg/time/#Parse
https://yourbasic.org/golang/format-parse-string-time-date-example/
"(This date is easier to remember when written as 01/02 03:04:05PM ‘06 -0700.)"

File: api/handlers_spaces.go (hunk 6 above)
Comment (damnski) [reply api/handlers_spaces.go]: done

File: api/routes.go (hunk 4 above)
Comment (damnski) [reply api/routes.go]: gotcha

File: api/handlers_spaces.go (hunk 5 above)
Comment (damnski) [reply api/handlers_spaces.go]: cleaned up

**Comment by damnski** [comment]
updated readme

File: api/handlers_spaces.go [hunk 7]
```diff
@@ -19,8 +19,19 @@ import (
 // it pulls data from the start of the month until now.
 func (s *server) SpaceGetHandler(w http.ResponseWriter, r *http.Request) {
 	w = LogWriter{w}
+
+	// loop thru given API input vars in debug
 	vars := mux.Vars(r)
+	for k, v := range mux.Vars(r) {
+		log.Debugf("key=%v, value=%v", k, v)
+	}
```
Comment (fishnix) [inline api/handlers_spaces.go]: range over `vars` instead of `mux.Vars(r)`

File: api/handlers_spaces.go [hunk 8]
```diff
@@ -37,15 +48,68 @@ func (s *server) SpaceGetHandler(w http.ResponseWriter, r *http.Request) {
...
+		ecleansedTime = fmt.Sprintf("%d-%02d-%02d", y, m, d)
+		scleansedTime = fmt.Sprintf("%d-%02d-01", y, m)
+		timeValidity = true
+	} else {
+		// working vars for time verification
+		var sTime string
+		var eTime string
+		var badTime bool
+
+		// debugging verbosity
```
Comment (fishnix) [inline api/handlers_spaces.go]: ✂️

File: api/handlers_spaces.go [hunk 9]
```diff
@@ -37,15 +48,68 @@ func (s *server) SpaceGetHandler(w http.ResponseWriter, r *http.Request) {
...
+		log.Debugf("startTime: %s", startTime)
+		log.Debugf("endTime: %s", endTime)
```
Comment (fishnix) [inline api/handlers_spaces.go]: combine these into one statement

File: api/handlers_spaces.go [hunk 10]
```diff
@@ -37,15 +48,68 @@ func (s *server) SpaceGetHandler(w http.ResponseWriter, r *http.Request) {
 	}
 	log.Debugf("found cost explorer result cache %+v", *resultCache)
 
-	spaceID := vars["space"]
-	log.Debugf("getting costs for space %s", spaceID)
+	// vars for checking input times parse and are valid
+	timeValidity := false
+	var scleansedTime string
+	var ecleansedTime string
```
Comment (fishnix) [inline api/handlers_spaces.go]: lets call these `start` and `end`?

File: api/handlers_spaces.go [hunk 11]
```diff
@@ -37,15 +48,68 @@ func (s *server) SpaceGetHandler(w http.ResponseWriter, r *http.Request) {
...
-	// if it's the first day of the month, get todays usage thus far
-	if d == 1 {
-		d = 2
+		// set the time parse layout and do parse
+		layoutISO := "2006-01-02"
+		start, err := time.Parse(layoutISO, startTime)
+		if err != nil {
+			msg := fmt.Sprintf("error parsing startTime from input: %s\n", err)
+			handleError(w, apierror.New(apierror.ErrBadRequest, msg, nil))
+			badTime = true
```
Comment (fishnix) [inline api/handlers_spaces.go]: ~`badtime = true`~ just `return`

File: api/handlers_spaces.go [hunk 12]
```diff
@@ -37,15 +48,68 @@ func (s *server) SpaceGetHandler(w http.ResponseWriter, r *http.Request) {
...
+
+		// debugging verbosity
+		log.Debugf("startTime: %s", startTime)
+		log.Debugf("endTime: %s", endTime)
 
-	// if it's the first day of the month, get todays usage thus far
-	if d == 1 {
-		d = 2
+		// set the time parse layout and do parse
+		layoutISO := "2006-01-02"
```
Comment (fishnix) [inline api/handlers_spaces.go]: i don't think you need this variable, the compiler probably optimizes it, but I would just put `"2006-01-02"` in the calls to `time.Parse()`

File: api/handlers_spaces.go [hunk 13]
```diff
@@ -37,15 +48,68 @@ func (s *server) SpaceGetHandler(w http.ResponseWriter, r *http.Request) {
...
+		start, err := time.Parse(layoutISO, startTime)
+		if err != nil {
+			msg := fmt.Sprintf("error parsing startTime from input: %s\n", err)
```
Comment (fishnix) [inline api/handlers_spaces.go]: `error parsing StartTime from input....`

File: api/handlers_spaces.go [hunk 14]
```diff
@@ -37,15 +48,68 @@ func (s *server) SpaceGetHandler(w http.ResponseWriter, r *http.Request) {
...
+			handleError(w, apierror.New(apierror.ErrBadRequest, msg, nil))
+			badTime = true
+		}
+		end, err := time.Parse(layoutISO, endTime)
+		if err != nil {
+			msg := fmt.Sprintf("error parsing endTime from input: %s\n", err)
+			handleError(w, apierror.New(apierror.ErrBadRequest, msg, nil))
+			badTime = true
+		}
```
Comment (fishnix) [inline api/handlers_spaces.go]: same minor things for this

File: api/handlers_spaces.go [hunk 15]
```diff
@@ -37,15 +48,68 @@ func (s *server) SpaceGetHandler(w http.ResponseWriter, r *http.Request) {
...
+
+		// convert times to number string yyMMdd so we can do a math comparison
+		sTime = fmt.Sprint(start.Format("20060102"))
+		eTime = fmt.Sprint(end.Format("20060102"))
+		log.Debugf("startYYYYMMDD: %s", sTime)
+		log.Debugf("endYYYYMMDD: %s", eTime)
```
Comment (fishnix) [inline api/handlers_spaces.go]: you can get rid of all of this (and below) and use https://golang.org/pkg/time/#Time.After

File: api/handlers_spaces.go [hunk 16]
```diff
@@ -37,15 +48,68 @@ func (s *server) SpaceGetHandler(w http.ResponseWriter, r *http.Request) {
...
+
+		// if time on the API input is already borked, don't continue
+		if badTime == false {
```
Comment (fishnix) [inline api/handlers_spaces.go]: ✂️ abandon ship above with return

File: api/handlers_spaces.go [hunk 17]
```diff
@@ -108,43 +172,54 @@ func (s *server) SpaceGetHandler(w http.ResponseWriter, r *http.Request) {
...
-	// the object is not found in the cache, call AWS cost-explorer and set cache
-	var out []*costexplorer.ResultByTime
-	c, expire, ok := resultCache.GetWithExpiration(spaceID)
-	if !ok || c == nil {
-		log.Debugf("cache empty for org, space: %s, %s, calling cost-explorer", Org, spaceID)
-		// call cost-explorer
-		var err error
-		out, err = ceService.GetCostAndUsage(r.Context(), &input)
+	// if time did not parse correctly, don't call AWS cost explorer
+	if timeValidity == true {
```
Comment (fishnix) [inline api/handlers_spaces.go]: you probably don't need this anymore, you can just use `start` and `end`

File: api/handlers_spaces.go [hunk 18]
```diff
@@ -108,43 +172,54 @@ func (s *server) SpaceGetHandler(w http.ResponseWriter, r *http.Request) {
...
+		// create a cacheKey more unique than spaceID for managing cache objects.
+		// Since we will accept date-range cost exploring, concatenate the spaceID
+		// and the start and end time so we can cache each time-based result
+		var cacheKey string
+		cacheKey = fmt.Sprintf("%s_%s_%s", spaceID, startTime, endTime)
```
Comment (fishnix) [inline api/handlers_spaces.go]: 😿

File: api/handlers_spaces.go (hunk 12 above)
Comment (damnski) [reply api/handlers_spaces.go]: It's DRYish - layoutISO is used twice.

Comment (fishnix) [reply api/handlers_spaces.go]: d its not really a useful variable - i think its also more clear to see the format you expect in the call to Parse()

File: api/handlers_spaces.go (hunk 10 above)
Comment (damnski) [reply api/handlers_spaces.go]: done.  Had to wrangle time.Time into string afterall.

File: api/handlers_spaces.go (hunk 11 above)
Comment (damnski) [reply api/handlers_spaces.go]: done

File: api/handlers_spaces.go (hunk 13 above)
Comment (damnski) [reply api/handlers_spaces.go]: ok

File: api/handlers_spaces.go (hunk 15 above)
Comment (damnski) [reply api/handlers_spaces.go]: nice - timeValidity = eTmp.After(sTmp)

**Comment by damnski** [comment]
Thanks for feedback

File: README.md [hunk 19]
```diff
@@ -10,6 +10,7 @@ GET /v1/cost/version
 GET /v1/cost/metrics
 
 GET /v1/cost/{account}/spaces/{spaceid}
+GET /v1/cost/{account}/spaces/{spaceid}?StartTime=2019-10-01&EndTime=2019-10-30
```
Comment (fishnix) [inline README.md]: `GET /v1/cost/{account}/spaces/{spaceid}[?StartTime=2019-10-01&EndTime=2019-10-30]`

File: api/handlers_spaces.go [hunk 20]
```diff
@@ -15,12 +15,60 @@ import (
 	log "github.com/sirupsen/logrus"
 )
 
+// getTimeDefault returns time range from beginning of month to day-of-month now
+func getTimeDefault() (string, string, bool) {
```
Comment (fishnix) [inline api/handlers_spaces.go]: the `bool` is not used in your return

File: api/handlers_spaces.go [hunk 21]
```diff
@@ -15,12 +15,60 @@ import (
...
+	log.Debug("no start or end time given on API input, assigning defaults")
```
Comment (fishnix) [inline api/handlers_spaces.go]: this doesn't make sense in this context actually - move the log message out to where it's called i think

File: api/handlers_spaces.go [hunk 22]
```diff
@@ -15,12 +15,60 @@ import (
...
+	// if it's the first day of the month, get today's usage thus far
+	y, m, d := time.Now().Date()
+	if d == 1 {
+		d = 3
+	}
+	return fmt.Sprintf("%d-%02d-01", y, m), fmt.Sprintf("%d-%02d-%02d", y, m, d), true
+}
+
+// timeAPI returns time parsed from API input
+func getTimeAPI(startTime, endTime string) (string, string, bool, error) {
```
Comment (fishnix) [inline api/handlers_spaces.go]: 👀

File: api/handlers_spaces.go [hunk 23]
```diff
@@ -15,12 +15,60 @@ import (
...
+	log.Debug("no start or end time given on API input, assigning defaults")
+	// if it's the first day of the month, get today's usage thus far
+	y, m, d := time.Now().Date()
+	if d == 1 {
+		d = 3
+	}
+	return fmt.Sprintf("%d-%02d-01", y, m), fmt.Sprintf("%d-%02d-%02d", y, m, d), true
+}
+
+// timeAPI returns time parsed from API input
```
Comment (fishnix) [inline api/handlers_spaces.go]: 👀

File: api/handlers_spaces.go [hunk 24]
```diff
@@ -37,15 +85,45 @@ func (s *server) SpaceGetHandler(w http.ResponseWriter, r *http.Request) {
...
+					w.WriteHeader(http.StatusBadRequest)
+				case apierror.ErrLimitExceeded:
+					w.WriteHeader(http.StatusTooManyRequests)
+				default:
+					w.WriteHeader(http.StatusInternalServerError)
+				}
+				w.Write([]byte(aerr.Message))
+			} else {
+				w.WriteHeader(http.StatusInternalServerError)
+				w.Write([]byte(err.Error()))
```
Comment (fishnix) [inline api/handlers_spaces.go]: `handleError()` does all of this already, I think you just need to call it here

https://github.com/YaleSpinup/cost-api/blob/master/api/handlers.go#L48

File: api/handlers_spaces.go [hunk 25]
```diff
@@ -108,43 +186,54 @@ func (s *server) SpaceGetHandler(w http.ResponseWriter, r *http.Request) {
...
-	// the object is not found in the cache, call AWS cost-explorer and set cache
-	var out []*costexplorer.ResultByTime
-	c, expire, ok := resultCache.GetWithExpiration(spaceID)
-	if !ok || c == nil {
-		log.Debugf("cache empty for org, space: %s, %s, calling cost-explorer", Org, spaceID)
-		// call cost-explorer
-		var err error
-		out, err = ceService.GetCostAndUsage(r.Context(), &input)
+	// if time did not parse correctly, don't call AWS cost explorer
+	if timeValidity == true {
```
Comment (fishnix) [inline api/handlers_spaces.go]: i still don't think we need this boolean?  (also `if timeValidity {`)

File: api/handlers_spaces_test.go [hunk 26]
```diff
@@ -0,0 +1,130 @@
...
+// getTimeDefault
+func TestGetTimeDefault(t *testing.T) {
+	// use defaults derived in code
+	startTime, endTime, outBool := getTimeDefault()
+	if outBool {
+		t.Logf("bool of getTimeDefault output: %v\n", outBool)
+	}
+	if outBool == false {
+		t.Errorf("bool of getTimeDefault false, expected true: %v\n", outBool)
+	}
```
Comment (fishnix) [inline api/handlers_spaces_test.go]: this will only ever be `true` from that function

File: api/handlers_spaces_test.go [hunk 27]
```diff
@@ -0,0 +1,130 @@
+package api
+
+import (
+	"fmt"
+	"testing"
+	"time"
+)
+
+// getTimeDefault
```
Comment (fishnix) [inline api/handlers_spaces_test.go]: 👀

File: api/handlers_spaces_test.go [hunk 28]
```diff
@@ -0,0 +1,130 @@
...
+	}
+	if endTime != eTime {
+		t.Logf("negative test eTime: %s does not match: %s", eTime, endTime)
+	} else {
+		t.Errorf("got unexpected eTime: %s\n", eTime)
+	}
+
+}
+
+// getTimeAPI
```
Comment (fishnix) [inline api/handlers_spaces_test.go]: 👀

File: api/handlers_spaces.go (hunk 23 above)
Comment (damnski) [reply api/handlers_spaces.go]: fixed up the name

File: api/handlers_spaces.go (hunk 21 above)
Comment (damnski) [reply api/handlers_spaces.go]: moved

File: api/handlers_spaces.go (hunk 20 above)
Comment (damnski) [reply api/handlers_spaces.go]: removed

File: README.md (hunk 19 above)
Comment (damnski) [reply README.md]: updated

File: api/handlers_spaces_test.go (hunk 26 above)
Comment (damnski) [reply api/handlers_spaces_test.go]: removed timeValidity, refactored to using return errors down to base function SpaceGetHandler

File: api/handlers_spaces.go (hunk 25 above)
Comment (damnski) [reply api/handlers_spaces.go]: gone

File: api/handlers_spaces_test.go (hunk 27 above)
Comment (damnski) [reply api/handlers_spaces_test.go]: gone

File: api/handlers_spaces_test.go (hunk 28 above)
Comment (damnski) [reply api/handlers_spaces_test.go]: gone

**Pull request merged**